google-cloud-clients/google-cloud-firestore/src/main/java/com/google/cloud/firestore/FirestoreOptions.java

```diff
@@ -216,8 +221,74 @@ public FirestoreOptions build() {
         }
       }
 
+      // Override credentials and channel provider if we are using the emulator.
+      String emulatorHost = System.getenv(FIRESTORE_EMULATOR_SYSTEM_PROPERTY);
+      if (emulatorHost != null) {
+        String hostUrlString = "http://" + emulatorHost;
+
+        // Try creating a host in order to validate that the host name is valid.
+        try {
+          URI.create(hostUrlString);
+        } catch (Exception e) {
+          throw new IllegalArgumentException(
+              "Value: '"
+                  + emulatorHost
+                  + "' for property "
+                  + FIRESTORE_EMULATOR_SYSTEM_PROPERTY
+                  + " is not a valid host",
+              e);
+        }
+
+        // Setting the channel provider to use plaintext turns off SSL.
+        this.setChannelProvider(
+            InstantiatingGrpcChannelProvider.newBuilder()
+                .setEndpoint(emulatorHost)
+                .setChannelConfigurator(
+                    new ApiFunction<ManagedChannelBuilder, ManagedChannelBuilder>() {
+                      @Override
+                      public ManagedChannelBuilder apply(ManagedChannelBuilder input) {
+                        input.usePlaintext();
+                        return input;
+                      }
+                    })
+                .build());
+        this.setCredentials(new FakeCredentials());
+      }
+
       return new FirestoreOptions(this);
     }
+
+    public class FakeCredentials extends Credentials {
+      private Map<String, List<String>> headers;
```

Comment:
```
private static final Map<String, List<String>> HEADERS = ImmutableMap.of("Authorization", ImmutableList.of("Bearer owner"))
```

we have Guava around, might as well use it

Reply:
I initially tried that when I started, but ran into the error
```
Error:(263, 72) java: incompatible types: com.google.common.collect.ImmutableMap<java.lang.String,com.google.common.collect.ImmutableList<java.lang.String>> cannot be converted to java.util.Map<java.lang.String,java.util.List<java.lang.String>>
```

Reply:
how about
```
ImmutableMap.of("Authorization", Arrays.asList("Bearer owner"))
```
?